### PR TITLE
Task functionality for both the Home Screen and the Detail Screen

### PR DESCRIPTION
**Pull request by MKiperszmid**
Preview of how the Task works

### CODE REVIEW

File: app/src/main/java/com/mk/tasky/agenda/data/AgendaRepositoryImpl.kt [hunk 1]
```diff
@@ -70,22 +70,26 @@ class AgendaRepositoryImpl(
...
+            emit(Agenda(reminders = localReminders, tasks = localTasks, events = emptyList()))
             if (forceRemote) {
                 getAgendaRemotely(date).onSuccess { response ->
                     supervisorScope {
                         response.reminders.map {
                             launch { dao.insertReminder(it.toDomain().toEntity()) }
                         }.forEach { it.join() }
+                        response.tasks.map {
+                            launch { dao.insertTask(it.toDomain().toEntity()) }
+                        }.forEach { it.join() }
```
Comment (philipplackner) [inline app/src/main/java/com/mk/tasky/agenda/data/AgendaRepositoryImpl.kt]: Ideally, this should use a separate supervisorScope, else inserting the tasks will wait until all reminders are inserted

File: app/src/main/java/com/mk/tasky/agenda/presentation/detail/task/DetailTaskViewModel.kt [hunk 2]
```diff
@@ -5,13 +5,20 @@ import androidx.compose.runtime.mutableStateOf
...
 import javax.inject.Inject
 
+@HiltViewModel
 class DetailTaskViewModel @Inject constructor(
-    savedStateHandle: SavedStateHandle
+    savedStateHandle: SavedStateHandle,
+    private val taskUseCases: TaskUseCases
 ) : ViewModel() {
     var state by mutableStateOf(DetailTaskState())
         private set
```
Comment (philipplackner) [inline app/src/main/java/com/mk/tasky/agenda/presentation/detail/task/DetailTaskViewModel.kt:24]: BTW, not sure if you want to do that already or do it at the end, but right now your app wouldn't survive process death

File: app/src/main/java/com/mk/tasky/agenda/presentation/home/AgendaItem.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import java.time.LocalDateTime
+
+data class AgendaItem(
+    val id: String,
+    val title: String,
+    val isDone: Boolean,
+    val description: String,
+    val type: HomeAgendaType,
+    val firstDatetime: LocalDateTime,
+    val secondDatetime: LocalDateTime? = null
```
Comment (philipplackner) [inline app/src/main/java/com/mk/tasky/agenda/presentation/home/AgendaItem.kt]: Hmm, I'm not sure about if it's a good idea to model it that way. This could make it a pain to extend in case there are more `HomeAgendaType`'s with individual behavior (such as `Event` which will require individual things like attendees and photos). Also, only tasks can be marked as done while your way of modeling it will add such a boolean to all types.

Why not model it somehow like this (this is from my project)? If you want to use the same name for the fields (like `id` instead of `eventId`), you'd need to make the child classes normal ones instead of data classes (but losing the advantages of data classes, which is why I'm fine with using different names)
```
sealed class AgendaItem(
    val id: String,
    val title: String,
    val description: String?,
    val sortDate: ZonedDateTime,
    val remindAt: ZonedDateTime,
    val reminderType: ReminderType
) {
    data class Event(
        val eventId: String,
        val eventTitle: String,
        val eventDescription: String?,
        val from: ZonedDateTime,
        val to: ZonedDateTime,
        val photos: List<EventPhoto>,
        val attendees: List<Attendee>,
        val isUserEventCreator: Boolean,
        val host: String?,
        val remindAtTime: ZonedDateTime,
        val eventReminderType: ReminderType
    ) : AgendaItem(eventId, eventTitle, eventDescription, from, remindAtTime, eventReminderType)

    data class Task(
        val taskId: String,
        val taskTitle: String,
        val taskDescription: String?,
        val time: ZonedDateTime,
        val isDone: Boolean,
        val remindAtTime: ZonedDateTime,
        val taskReminderType: ReminderType
    ) : AgendaItem(taskId, taskTitle, taskDescription, time, remindAtTime, taskReminderType)

    data class Reminder(
        val reminderId: String,
        val reminderTitle: String,
        val reminderDescription: String?,
        val time: ZonedDateTime,
        val remindAtTime: ZonedDateTime,
        val reminderReminderType: ReminderType
    ) : AgendaItem(reminderId, reminderTitle, reminderDescription, time, remindAtTime, reminderReminderType)
}
```

File: app/src/main/java/com/mk/tasky/agenda/presentation/home/AgendaItemMapper.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package com.mk.tasky.agenda.presentation.home
+
+import com.mk.tasky.agenda.domain.model.Agenda
+import com.mk.tasky.agenda.domain.model.Reminder
+import com.mk.tasky.agenda.domain.model.Task
+
+// TODO: Add Event
+fun Agenda.toAgendaItem(): List<AgendaItem> {
+    return this.reminders.map { it.toAgendaItem() } + this.tasks.map { it.toAgendaItem() }
```
Comment (philipplackner) [inline app/src/main/java/com/mk/tasky/agenda/presentation/home/AgendaItemMapper.kt]: Is this only related to presentation? Might rather put this in domain or data directly and expose a list of `AgendaItems` to presentation

**Review by philipplackner — APPROVED** [review]

File: app/src/main/java/com/mk/tasky/agenda/presentation/detail/task/DetailTaskViewModel.kt (hunk 2 above)
Comment (MKiperszmid) [reply app/src/main/java/com/mk/tasky/agenda/presentation/detail/task/DetailTaskViewModel.kt:24]: I'm going to do it at the end of the whole app, as the "final touch"

File: app/src/main/java/com/mk/tasky/agenda/presentation/home/AgendaItem.kt (hunk 3 above)
Comment (MKiperszmid) [reply app/src/main/java/com/mk/tasky/agenda/presentation/home/AgendaItem.kt]: Ohh that's interesting
Watchout cause the next commits are going to be big then since we are removing the Task and Reminder as we know it, and it's all going to inherit from AgendaItem

File: app/src/main/java/com/mk/tasky/agenda/presentation/home/AgendaItemMapper.kt (hunk 4 above)
Comment (MKiperszmid) [reply app/src/main/java/com/mk/tasky/agenda/presentation/home/AgendaItemMapper.kt]: Removed it after the migration to AgendaItem, since it didn't made much sense

File: app/src/main/java/com/mk/tasky/agenda/data/AgendaRepositoryImpl.kt [hunk 5]
```diff
@@ -70,25 +71,23 @@ class AgendaRepositoryImpl(
...
+            emit(Agenda(localReminders + localTasks))
             if (forceRemote) {
                 getAgendaRemotely(date).onSuccess { response ->
                     supervisorScope {
                         response.reminders.map {
                             launch { dao.insertReminder(it.toDomain().toEntity()) }
                         }.forEach { it.join() }
+                        response.tasks.map {
+                            launch { dao.insertTask(it.toDomain().toEntity()) }
+                        }.forEach { it.join() }
```
Comment (philipplackner) [inline app/src/main/java/com/mk/tasky/agenda/data/AgendaRepositoryImpl.kt]: I think I commented on something like this the last time right? Here, inserting the tasks would start after all reminders have been inserted which can all happen in parallel. Therefore, either use a separate scope or just save references to the jobs and join them all together at the end of the scope.

File: app/src/main/java/com/mk/tasky/agenda/domain/model/AgendaItem.kt [hunk 6]
```diff
@@ -0,0 +1,40 @@
+package com.mk.tasky.agenda.domain.model
+
+import java.time.LocalDateTime
+
+sealed class AgendaItem(
+    val id: String,
+    val title: String,
+    val description: String,
+    val firstDatetime: LocalDateTime,
```
Comment (philipplackner) [inline app/src/main/java/com/mk/tasky/agenda/domain/model/AgendaItem.kt]: I find the name `firstDatetime` says quite little about what it represents. Would it be clear to you if you'd start working on this project if someone else wrote this?
But I also struggled to find a good name for this, which is why I came up with `sortDate`, pretty much the date we want to sort the AgendaItems by on the AgendaScreen. Not ideal either, but what about just calling it `time` or `dateTime` and then Events just have another field called `toTime`? 🤔

**Review by philipplackner — APPROVED** [review]

File: app/src/main/java/com/mk/tasky/agenda/data/AgendaRepositoryImpl.kt (hunk 5 above)
Comment (MKiperszmid) [reply app/src/main/java/com/mk/tasky/agenda/data/AgendaRepositoryImpl.kt]: Will do the last one. Grab all the jobs and join them
Thanks!

File: app/src/main/java/com/mk/tasky/agenda/domain/model/AgendaItem.kt (hunk 6 above)
Comment (MKiperszmid) [reply app/src/main/java/com/mk/tasky/agenda/domain/model/AgendaItem.kt]: Renamed to time
